packages/cdktf/lib/terraform-asset.ts

```diff
@@ -63,12 +64,18 @@ export class TerraformAsset extends Resource {
         `TerraformAsset ${id} expects path to be a file, a directory was passed: '${config.path}'`
       );
     }
+
+    if (this.stack.node.root instanceof App) {
+      this.copyAsset(this.stack.node.root.manifest);
+    } else {
+      throw new Error("Could not copy asset because no root app was found");
+    }
```

Comment:
I believe that's called too early? Right now, that's called when `app.synth()` is called. With this change, it would be on resource construction.

Reply:
But wouldn't you have an app in place whenever you instantiate a stack (passing the app) which instantiates the resources (passing the stack which has an app?)

Reply:
While the app creates the [outdir](https://github.com/hashicorp/terraform-cdk/blob/a5eeb9e5901d5918407d6f0246f383c746391585/packages/cdktf/lib/app.ts#L84) in the constructor, the stack directories are created in the [StackSynthesizer](https://github.com/hashicorp/terraform-cdk/blob/a5eeb9e5901d5918407d6f0246f383c746391585/packages/cdktf/lib/synthesize/synthesizer.ts#L29).

Plus, there might be adjustments happening in Aspects (e.g. changing code - thinking of Lambda stubs).

Reply:
✅ Changed the behavior to copy files on synth (added support for a custom synth function)